### PR TITLE
Add `findall` method

### LINKED ISSUE #217: Implement `findall`
We don't have this function, but we might as well.
We might want to not return a `Vector` of results but a lazy iterator though (what's up with Base needlessly materializing everything?!)

### PR DESCRIPTION
**Pull request by CiaranOMara**
I reckon this PR fulfils the `findall` method requirements. 

The `findall` method delegates to a `Search` iterator in the implementation. Let me know whether this approach suits and I'll add some docstrings.

Closes #217.

### CODE REVIEW

File: src/BioSequences.jl [hunk 1]
```diff
@@ -227,4 +227,46 @@ include("search/ApproxSearchQuery.jl")
...
+
+const HasRangeEltype = Union{<:ExactSearchQuery, <:ApproximateSearchQuery, <:Regex}
+
+Base.eltype(::Type{Search{Q,I}}) where {Q<:HasRangeEltype,I} = UnitRange{Int}
+Base.IteratorEltype(::Type{Search{Q,I}}) where {Q<:HasRangeEltype,I} = Base.IteratorEltype(UnitRange{Int})
+
+Base.eltype(::Type{<:Search}) = Int
+Base.IteratorEltype(::Type{<:Search}) = Base.IteratorEltype(Int)
+
+Base.IteratorSize(::Type{<:Search}) = Base.SizeUnknown()
```
Comment (jakobnissen) [inline src/BioSequences.jl:260]:
```suggestion
Base.eltype(::Type{Search{<:Q}}) where {Q<:HasRangeEltype} = UnitRange{Int}
Base.eltype(::Type{<:Search}) = Int
Base.IteratorSize(::Type{<:Search}) = Base.SizeUnknown()
```

Comment (jakobnissen) [reply src/BioSequences.jl:260]: `Base.IteratorSize` just needs to be `HasEltype()`, but this is already the default.

File: src/BioSequences.jl [hunk 2]
```diff
@@ -227,4 +227,46 @@ include("search/ApproxSearchQuery.jl")
...
+
+function Base.findall(pat, seq::BioSequence, start::Integer=firstindex(seq); overlap::Bool = default_overlap, stop::Integer = lastindex(seq))
```
Comment (jakobnissen) [inline src/BioSequences.jl:262]: I'm ambivalent about this signature. Why is start part of the signature but stop isn't? Might be good though, just not sure.

File: src/BioSequences.jl [hunk 3]
```diff
@@ -227,4 +227,46 @@ include("search/ApproxSearchQuery.jl")
...
+
+struct Search{Q,I}
+    query::Q
+    itr::I
+    start::Integer
+    stop::Integer
+    overlap::Bool
+end
+
+const default_overlap = true
```
Comment (jakobnissen) [inline src/BioSequences.jl:240]:
```suggestion
const DEFAULT_OVERLAP = true
```

File: src/BioSequences.jl (hunk 2 above)
Comment (jakobnissen) [inline src/BioSequences.jl:262]:
```suggestion
function Base.findall(pat, seq::BioSequence, start::Integer=firstindex(seq); overlap::Bool = DEFAULT_OVERLAP, stop::Integer = lastindex(seq))
```

File: src/BioSequences.jl [hunk 4]
```diff
@@ -227,4 +227,46 @@ include("search/ApproxSearchQuery.jl")
 include("search/re.jl")
 include("search/pwm.jl")
 
+
+
+struct Search{Q,I}
+    query::Q
+    itr::I
+    start::Integer
+    stop::Integer
```
Comment (jakobnissen) [inline src/BioSequences.jl:236]:
```suggestion
    start::Int
    stop::Int
```

**Review by jakobnissen — COMMENTED** [review]
Great, thanks for contributing. Not sure about the signature though, it doesn't match `Base`. What signature would you prefer?

File: src/BioSequences.jl (hunk 2 above)
Comment (jakobnissen) [reply src/BioSequences.jl:262]: I think I would prefer having the same signature as `Base.findall(::AbstractString, ::AbstractString; overlap=true)` - though `Base` has overlap default to `false`, that seems strange to me.

If the user wants to limit the start and stop, they can create a view of the sequence.

File: src/BioSequences.jl (hunk 1 above)
Comment (CiaranOMara) [reply src/BioSequences.jl]: @jakobnissen, do we need the `<:Q`?
```julia
Base.eltype(::Type{<:Search{<:Q}}) where {Q<:HasRangeEltype} = UnitRange{Int}
vs
Base.eltype(::Type{<:Search{Q}}) where {Q<:HasRangeEltype} = UnitRange{Int}
```

Comment (jakobnissen) [reply src/BioSequences.jl]: Ah no, you are right, we don't

**Comment by CiaranOMara** [comment]
This PR is ready for another review pass.

I've added a convenience method that uses a view to limit the search range and map indices back to a parent. If it's not worthwhile, it can be removed.
https://github.com/BioJulia/BioSequences.jl/blob/fd0f57fa26abe0beffc05aa9bbadad0bac426e6c/src/BioSequences.jl#L259-L263

Also, is the PR code positioned where you'd like it to be within the project?

File: src/BioSequences.jl (hunk 2 above)
Comment (CiaranOMara) [reply src/BioSequences.jl]: An overlap default of `true` makes more sense to me in the context of BioSequences. But I'm okay changing to be consistent with `Base` if that's the consensus. However, setting the default to `false` may introduce a risk of missing important sequences.

Comment (jakobnissen) [reply src/BioSequences.jl]: I agree - `true` is the most sensible default here 👍

**Review by jakobnissen — APPROVED** [review]

**Comment by jakobnissen** [comment]
I think everything looks good now.

**Comment by SabrinaJaye** [comment]
Looking good! Did you want to add some docstrings and or a paragraph in the search section of the manual? No immediate pressure as they can always be added before a release.

**Comment by CiaranOMara** [comment]
I reckon I can add some docstrings to this over the next couple of days.

**Comment by jakobnissen** [comment]
I'll just merge this, and then we can always add some docs later.

**Pull request merged**